### PR TITLE
Dependencies injected so tests used mocked backendSrv

### LINKED ISSUE #20531: DataSourceSettingsPage: state set in promise after component unmounted
Should either use Redux or cancelable promise. I think at this point, @dprokop would recommend Redux.

```
backend.js:6 Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in the componentWillUnmount method.
    in DataSourceSettingsPage (created by ConnectFunction)
    in ConnectFunction (created by HotExportedConnect(DataSourceSettingsPage))
    in AppContainer (created by HotExportedConnect(DataSourceSettingsPage))
    in HotExportedConnect(DataSourceSettingsPage) (created by LoadableComponent)
    in LoadableComponent (created by Component)
    in Component (created by Component)
    in ThemeProvider (created by Component)
    in Component (created by ConfigProvider)
    in ConfigProvider
    in ErrorBoundary (created by ErrorBoundaryAlert)
    in ErrorBoundaryAlert
    in Provider
```

### PR DESCRIPTION
**Pull request by shavonn**
Fixes #20531

### CODE REVIEW

**Comment by marefr** [comment]
@shavonn #20531 is already closed for 6.6.0-stable milestone. Is this a fix of that same issue? If so, would suggest reopen issue and move to 6.6.1 milestone.

**Comment by shavonn** [comment]
> @shavonn #20531 is already closed for 6.6.0-stable milestone. Is this a fix of that same issue? If so, would suggest reopen issue and move to 6.6.1 milestone.

@marefr Yes, this is a fix for the same issue. Hugoh said because the test used the real backendSrv, it wasn't included in 6.6 so to update this so it would go into 6.6.1. I'll change the tag on the issue. Do I need to do this on the other PR too?

**Comment by marefr** [comment]
@shavonn 
> Do I need to do this on the other PR too?

The PR closed for 6.6 stable? No

**Comment by shavonn** [comment]
> @shavonn
> 
> > Do I need to do this on the other PR too?
> 
> The PR closed for 6.6 stable? No
He had already updated it (because it didn't go in 6.6).

**Comment by hugohaggmark** [comment]
@marefr during the 6.6 stable release we decided to drop this PR because of the tests failing and we needed to rewrite them to make them work. So both this and the other PR should be re-targeted for 6.6.1 and still have `changelog`/`cherry-pick` needed label.

**Review by hugohaggmark — APPROVED** [review]
LGTM!

File: public/app/features/datasources/state/actions.ts [hunk 1]
```diff
@@ -67,17 +73,23 @@ export const initDataSourceSettings = (
   };
 };
 
-export const testDataSource = (dataSourceName: string): ThunkResult<void> => {
+export const testDataSource = (
+  dataSourceName: string,
+  dependencies: TestDataSourceDependencies = {
```
Comment (dprokop) [inline public/app/features/datasources/state/actions.ts]: I'm not sure changing this API is required. If I understand correctly the issue was that in the tests the services were not instantiated. I think mocking the required services as you did here https://github.com/grafana/grafana/blob/ee93fdff592e6c186a206b194297c3378670e897/public/app/features/datasources/state/actions.test.ts#L21 and then applying @grafana/runtime mock as we do for example here: https://github.com/grafana/grafana/blob/cf2cc713933599e7646416a56a665282c9d9e3bb/public/app/plugins/datasource/grafana-azure-monitor-datasource/app_insights/app_insights_datasource.test.ts#L6 should do the job.

Let me know @shavonn if that's unclear!

**Review by dprokop — CHANGES_REQUESTED** [review]
As described in my comment

File: public/app/features/datasources/state/actions.ts (hunk 1 above)
Comment (dprokop) [reply public/app/features/datasources/state/actions.ts]: @hugohaggmark FYI ^^

Comment (hugohaggmark) [reply public/app/features/datasources/state/actions.ts]: We did this pattern previously for InitDataSources so I think that's why Shavonn continued this way.

Comment (hugohaggmark) [reply public/app/features/datasources/state/actions.ts]: Makes it superclear what dependencies a thunk needs imo.

Comment (hugohaggmark) [reply public/app/features/datasources/state/actions.ts]: @dprokop ^^^

Comment (dprokop) [reply public/app/features/datasources/state/actions.ts]: @shavonn We have discussed above considerations with @hugohaggmark and agreed the implementation proposed by you is better than the one suggested by me. I have not noticed the defaults in the thunk definition at first, which led me to have second thoughts. Let's merge this as it is!

**Review by dprokop — APPROVED** [review]
Good stuff!

File: public/app/features/datasources/state/actions.ts (hunk 1 above)
Comment (shavonn) [reply public/app/features/datasources/state/actions.ts]: Thanks! Was out the last couple of days.

**Pull request merged**